### PR TITLE
Add support for types derived from supported BCL collections

### PR DESCRIPTION
**Pull request by layomia**
Fixes https://github.com/dotnet/corefx/issues/38767.
Fixes https://github.com/dotnet/corefx/issues/38521.

Supported in this PR: user-defined types that implement [natively supported collections](https://github.com/dotnet/corefx/issues/36643) that have "add" methods e.g. `IList<T>`, `Queue<T>`, `IDictionary`, `IList` e tc.

Not supported in this PR: user-defined types that implement natively supported collections that don't have "add" methods e.g. `IEnumerable<T>`, `ICollection`, `IReadOnyList<T>` and immutable collections.

### CODE REVIEW

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.AddProperty.cs [hunk 1]
```diff
@@ -67,7 +94,13 @@ private JsonPropertyInfo AddProperty(Type propertyType, PropertyInfo propertyInf
             return jsonInfo;
         }
 
-        internal static JsonPropertyInfo CreateProperty(Type declaredPropertyType, Type runtimePropertyType, PropertyInfo propertyInfo, Type parentClassType, JsonSerializerOptions options)
+        internal static JsonPropertyInfo CreateProperty(
+            Type declaredPropertyType,
+            Type runtimePropertyType,
+            Type implementedPropertyType,
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.AddProperty.cs]: How is implementedPropertyType different than runtimePropertyType? runtimePropertyType is intended to be determined at runtime...

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+        {
+            typeof(IReadOnlyDictionary<,>),
+            typeof(IReadOnlyCollection<>),
+            typeof(IReadOnlyList<>),
+            typeof(IEnumerable<>),
+        };
+
+        public static Type GetImplementedCollectionType(Type queryType)
+        {
+            Debug.Assert(queryType != null);
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs]: Could a check to queryType.IsAbstract and queryType.IsInterface help determine if a user-specified class should be used?

**Comment by danmosemsft** [comment]
At this point, is this required for 3.0? Is risk acceptable?

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.AddProperty.cs (hunk 1 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.AddProperty.cs]: I've removed the `TImplementedType` generic parameter because we don't need it.

The implementedPropertyType is one of the types that the query type derives from. We select it based on the best fit for our deserialization logic. If the query type is one which the serializer natively supports, the implemented type is the query type itself.
The implemented type is sometimes the same as the runtime property type e.g.

```C#
public class StringListWrapper : List<string> { }
```
Declared type: `typeof(StringListWrapper)`
Implemented type: `typeof(List<string>)`
Runtime type: `typeof(List<string>)`

In some cases the runtime property type is different:

```C#
public class GenericIListWrapper<T> : IList<T>
{
    private readonly List<T> _list = new List<T>();
    
    // IList<T> implementation
    // ...
}
```

Declared type: `typeof(GenericIListWrapper<T>)`
Implemented type: `typeof(IList<T>)`
Runtime type: `typeof(List<T>)`

In this example, this is because we treat IList<T> as List<T>:

Declared type: `typeof(IList<T>)`
Implemented type: `typeof(IList<T>)`
Runtime type: `typeof(List<T>)`

Keeping track of the implemented property type helps us support these derived types e.g here https://github.com/dotnet/corefx/blob/cb54f66c1e0f19d4b97fb19e5326a03d15415776/src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs#L207

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs (hunk 2 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs]: Yes, I've added this check.

**Comment by layomia** [comment]
> At this point, is this required for 3.0? Is risk acceptable?

Yes, it fixes some important issues raised by partners and community members, and it's pretty well tested.

https://github.com/dotnet/corefx/issues/38767.
https://github.com/dotnet/corefx/issues/38521.

**Comment by danmosemsft** [comment]
Ok.

**Comment by layomia** [comment]
cc @pranavkm

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/DefaultIDictionaryConverter.cs [hunk 3]
```diff
@@ -8,7 +8,7 @@ namespace System.Text.Json.Serialization.Converters
 {
     internal sealed class DefaultIDictionaryConverter : JsonDictionaryConverter
     {
-        public override IDictionary CreateFromDictionary(ref ReadStack state, IDictionary sourceDictionary, JsonSerializerOptions options)
+        public override object CreateFromDictionary(ref ReadStack state, IDictionary sourceDictionary, JsonSerializerOptions options)
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Serialization/Converters/DefaultIDictionaryConverter.cs]: Why the change to `object`? Don't all dictionaries implement `IDictionary` (non generic)?

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+
+            if (queryType.IsAbstract ||
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs]: Can we have a fast-path to exit if IEnumerable not implemented?

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+                    return derivedGeneric;
+                }
+            }
+
+            // Try non-generic interfaces without add methods
+            if (typeof(IEnumerable).IsAssignableFrom(queryType))
+            {
+                return typeof(IEnumerable);
+            }
+            else if (typeof(ICollection).IsAssignableFrom(queryType))
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs]: nit: don't need `else` here (or add `else` below to make it consistent)

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+                queryType.IsArray ||
+                queryType == typeof(object) ||
+                IsNativelySupportedCollection(queryType))
+            {
+                return queryType;
+            }
+
+            Type baseType = queryType.GetTypeInfo().BaseType;
+
+            // Check if the base type is a supported concrete collection.
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs]: Is it worthwhile to do this recursively? (any base class)?

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+            if (queryType == null)
+            {
+                return false;
+            }
+
+            if (queryType.IsGenericType)
+            {
+                Type typeDef = queryType.GetGenericTypeDefinition();
+
+                switch (typeDef.FullName)
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs]: It seems like this should be a hashtable\hashet since the list is getting long.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs (hunk 4 above)
Comment (steveharter) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs]: Or use GetClassType() instead?

File: src/System.Text.Json/tests/Serialization/TestClasses.NonGenericCollections.cs [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+            _list.Remove(value);
+        }
+
+        public void RemoveAt(int index)
+        {
+            _list.RemoveAt(index);
+        }
+    }
+
+    public class IDictionaryWrapper : IDictionary
```
Comment (steveharter) [inline src/System.Text.Json/tests/Serialization/TestClasses.NonGenericCollections.cs]: Hmm would `WrapperForIDictionary` be better since by beginning with `I` it seems like it should be an interface.

Comment (ericstj) [reply src/System.Text.Json/tests/Serialization/TestClasses.NonGenericCollections.cs]: Or just leave out I.

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/DefaultIDictionaryConverter.cs (hunk 3 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/DefaultIDictionaryConverter.cs]: Turns out that [`IDictionary<TKey, TValue`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.idictionary-2?view=netcore-3.0) does not. So we can't support (user-defined) types that implement only  `IDictionary<TKey, TValue>` with this converter mechanism.

I think it's better to change to object here than to implement a new converter mechanism for just these types, unless boxing to `object` is significantly worse than to `IDictionary` or `IDictionary<,>`.

In the meantime, I've added a comment to `JsonDictionaryConverter` stating that it's usage is generalized to deserializing (some) collections that store key-value pairs (not including `KeyValuePair<,>`) -- if it's okay to leave as object.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs (hunk 4 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs]: I've reordered to have a fast-path exit for types that don't implement `IEnumerable` and `string` types. We still need to check for `string` because it implements `IEnumerable`.

Removed the check for `object`.

`GetClassType()` calls this method - using it would be an unwanted circular dependency.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs (hunk 6 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs]: I don't think so. We'd still need to check for implemented interfaces, so not having any base class won't tell us much.

If we want to support detection of an implemented concrete class at a higher level than a types parent class, then we could recurse to the base type (if any) at the last line in this method. I'd hold of on this until there's a demonstrated user need though.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs (hunk 5 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs]: I think we do because we want to return the `queryType` below this if it doesn't implement `ICollection`.

Also, consistent with what?

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/DefaultIDictionaryConverter.cs (hunk 3 above)
Comment (steveharter) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/DefaultIDictionaryConverter.cs]: OK makes sense

**Comment by phizch** [comment]
This change makes it impossible to have a `JsonConverterFactory` distinguish between a BCL base class and a derived type.

Specialized collections can be necessary if the json has multiple ways of declaring a list, e.g.
``` json
{
  "List1" : 
  {
    "Count": 50,
    "Values" : [ "test", ... ]
  },
  "List2" : [ "test", ... ] 
}
``` 


``` CSharp
public class A
{
    public MyCollection List1 { get; set; }
    public List<string> List2 { get; set }
}
public class MyCollection : List<string> { }
public class MyConverterFactory : JsonConverterFactory
{
    public override bool CanConvert( Type typeToConvert )
    {
        return ( typeToConvert == typeof( MyCollection ) );
    }
    public override JsonConverter CreateConverter( Type typeToConvert, JsonSerializerOptions options )
    {
        // return the converter
        throw new NotImplementedException();
    }
}
```
From `JsonClassInfo.AddProperty`:
``` CSharp
    // Get implemented type, if applicable.
    // Will return the propertyType itself if it's a non-enumerable, string, or natively supported collection.
    Type implementedType = GetImplementedCollectionType(propertyType);
    if (implementedType != propertyType)
    {
        jsonInfo = CreateProperty(implementedType, implementedType, implementedType, null, typeof(object), options);
    }
    else
    {
        jsonInfo = CreateProperty(propertyType, propertyType, propertyType, propertyInfo, classType, options);
    }
```
`GetImplementedCollectionType(typeof(MyCollection))` will return `List<string>`, so CreateProperty will ask any factories to create converters for List<B>.

A solution could be to provide `CreateProperty` and on to `GetClassType` with both `implementedType` and `propertyType` so that it can send `propertyType` to any factories, while using the BCL type if no factory can produce a converter for the derived type. 

I don't know if I should have opened a new issue for this, but now at least you know..
If you'd like I can open a new issue and implement the changes.

**Comment by scalablecory** [comment]
@phizch this PR is closed and won't be tracked; please open a new issue for this. It's okay to reference this PR in that, though.

CC @ahsonkhan

**Comment by ahsonkhan** [comment]
> If you'd like I can open a new issue

Yes, please.
